Make DatadirTimeline a trait, implemented by LayeredTimeline.

hlinnaka opened the pull request:
Previously DatadirTimeline was a separate struct, and there was a 1:1
relationship between each DatadirTimeline and LayeredTimeline. That was
a bit awkward; whenever you created a timeline, you also needed to create
the DatadirTimeline wrapper around it, and if you only had a reference
to the LayeredTimeline, you would need to look up the corresponding
DatadirTimeline struct through tenant_mgr::get_local_timeline_with_load().
There were a couple of calls like that from LayeredTimeline itself.

Refactor DatadirTimeline, so that it's a trait, and mark LayeredTimeline
as implementing that trait. That way, there's only one object,
LayeredTimeline, and you can call both Timeline and DatadirTimeline
functions on that. You can now also call DatadirTimeline functions from
LayeredTimeline itself; there were a couple of places in LayeredTimeline
had to find its own DatadirTimeline struct by looking it up in tenant_mgr,
which was awkward.

I considered just moving all the functions from DatadirTimeline directly
to Timeline/LayeredTimeline, but I still like to have some separation.
Timeline provides a simple key-value API, and handles durably storing
key/value pairs, and branching. Whereas DatadirTimeline is stateless, and
provides an abstraction over the key-value store, to present an interface
with relations, databases, etc. Postgres concepts.

SomeoneToIgnore reviewed (commented):
I definitely like the way the PR moves to, especially that explicit repo knowledge is gone:

```rust
pub fn import_timeline_from_postgres_datadir<R: Repository>(
...
    tline: &mut DatadirTimeline<R>,
...
```

yet unfortunately, it gets masked as 

> `pub type TimelineImpl = <LayeredRepository as repository::Repository>::Timeline;`

, so the core types are still quite complicated.


I also find it a bit suspicious that `DatadirTimeline: Timeline` and does not need to have any its own methods implemented: to me, this looks like both traits will always be used on the same struct always, despite being separate.
Moreover, `DatadirTimeline` has methods for lower level (pages, relations, etc.) than `Timeline` that has  `checkpoint`, `get_ancestor_timeline_id`, etc.: it's odd to see the dependency direction placed as `DatadirTimeline: Timeline`.

Do we have a good confidence that both `Timeline` and `DatadirTimeline` traits are good API  borders that will be used with other kinds of storage systems?
If not, would it make sense to actually remove all traits entirely and create structs/enums only instead?
This way, if somebody would want to separate the entities, it has to think of data separation which is more important to deal with than public API via traits IMO.

hlinnaka commented:
Rebased over main.

This simplified the logical size calculation fast-path for branch creation, introduced in commit 28243d68e6. LayerTimeline can now access the ancestor's logical size directly, so it doesn't need the caller to pass it to it. I moved the fast-path to init_logical_size() function itself. It now checks if the ancestor's last LSN is the same as the branch point, i.e. if there haven't been any changes on the ancestor after the branch, and copies the size from there. An additional bonus is that the optimization will now work any time you have a branch of another branch, with no changes from the ancestor, not only at a create-branch command. @knizhnik

hlinnaka commented:
> I also find it a bit suspicious that `DatadirTimeline: Timeline` and does not need to have any its own methods implemented: to me, this looks like both traits will always be used on the same struct always, despite being separate.

Yes, indeed, they are always used together.

> Moreover, `DatadirTimeline` has methods for lower level (pages, relations, etc.) than `Timeline` that has `checkpoint`, `get_ancestor_timeline_id`, etc.: it's odd to see the dependency direction placed as `DatadirTimeline: Timeline`.
> 
> Do we have a good confidence that both `Timeline` and `DatadirTimeline` traits are good API borders that will be used with other kinds of storage systems? If not, would it make sense to actually remove all traits entirely and create structs/enums only instead? This way, if somebody would want to separate the entities, it has to think of data separation which is more important to deal with than public API via traits IMO.

The division is that `Timeline` represents a key-value store, with support for historical querying at a particular LSN, and branching. `DatadirTimeline` maps PostgreSQL data directory concepts like relations and pages to key-value pairs. So those pages and relations are the *higher* level concept, and keys and values are the lower level concept.

It's not perfect. I'm pretty happy with DatadirTimeline, it encapsulates everything related to Postgres datadir layout pretty well. Looking at `Repository/Timeline` traits, however, I think they expose many things that probably should be considered as implementation details, like `checkpoint`, `compaction_iteration`, `gc_iteration`, `get_remote_index`. If you implemented a different storage, you might not have the concepts of checkpointing, compaction or gc at all. They probably should be just removed from the trait.

So yeah, further refactoring and cleanup is surely in order. But I think this is a step in the right direction.

hlinnaka commented:
> > I also find it a bit suspicious that `DatadirTimeline: Timeline` and does not need to have any its own methods implemented: to me, this looks like both traits will always be used on the same struct always, despite being separate.
> 
> Yes, indeed, they are always used together.

One way to think about this is that if we started to support some other database systems than PostgreSQL on top of Neon, we would introduce a new trait like `DatadirTimeline`, something like `MysqlDatadirTimeline`. We might also need to do that for different Postgres versions, actually, if the new Postgres version makes changes to the data directory layout, like introduces new kinds of files that we need to store in the pageserver. But the code in `Timeline` would be unchanged.

It's not a perfect abstraction, we leak some details of WAL record types in the `Value` enum, in `ZenithWalRecord`, for example. We might need to refactor and tighten that up as part of the Postgres v15 support.

LizardWizzard reviewed (approved):
I can agree with both points :) I see the intention to keep the separation. It is a good thing because it allows to split the code by some logical aspect. Moving everything to Timeline leads to 3k+ lines in timeline.rs which makes it harder to navigate. I

> One way to think about this is that if we started to support some other database systems than PostgreSQL on top of Neon, we would introduce a new trait like DatadirTimeline, something like MysqlDatadirTimeline. 

Hypotetical MySQLDatadir cannot implement the same DatadirTimeline. Methods needed for MySQL for sure will be different. So I cannot imagine common API surface for them. Timeline trait can be shared in terms of its KV api, but Datadir impls wil be quite different. 

Regarding postgresql 15 it is hard to say before we know all the changes, what can be shared what cannot. It can be better to have different impls inside enum with different methods specific for each version. So we match against version and call what is needed. If we wont need different impls to support postgresql 15 IMO there is no need to keep it as a separate trait. 

So I agree that having traits that have only one impl is questionable. I'm not sure that traits we have provide good api boundaries. As @hlinnaka mentioned both Repository and Timeline are full of implementation details and I think this is mainly because there is only one impl. I agree that we need to clean it up if we want to keep the ability to relatively easy swap storage implementations. 

However I like the direction of this PR. It removes circular dependency between Timeline and DatadirTimeline, and makes code that uses both of them cleaner, so I propose to merge these changes and evolve on that further. Maybe during discussions for the RFC we'll come up with better alternative.

hlinnaka commented:
> I can agree with both points :) I see the intention to keep the separation. It is a good thing because it allows to split the code by some logical aspect. Moving everything to Timeline leads to 3k+ lines in timeline.rs which makes it harder to navigate. I
> 
> > One way to think about this is that if we started to support some other database systems than PostgreSQL on top of Neon, we would introduce a new trait like DatadirTimeline, something like MysqlDatadirTimeline.
> 
> Hypotetical MySQLDatadir cannot implement the same DatadirTimeline. Methods needed for MySQL for sure will be different. So I cannot imagine common API surface for them. Timeline trait can be shared in terms of its KV api, but Datadir impls wil be quite different.

Right, so the point is *not* that they would have a common API. The point is that there would be `trait PostgresDatadirTimeline` and also `trait MySqlDatadirTimeline`. Timeline can implement both, and if you want to store Postgres data in the underlying repository, you call the functions in `PostgresDatadirTimeline`, and if you want to store MySQL data, you call the functions in `MySqlDatadirTimeline`.

I added a comment on that to `DatadirTimeline`.

> Regarding postgresql 15 it is hard to say before we know all the changes, what can be shared what cannot. It can be better to have different impls inside enum with different methods specific for each version. So we match against version and call what is needed. If we wont need different impls to support postgresql 15 IMO there is no need to keep it as a separate trait.

Yeah, I'm not sure how exactly this will look like for v15. Where do we need to check and switch between versions? The differences might even cut through to the network API that page_service.rs provides.

> As @hlinnaka mentioned both Repository and Timeline are full of implementation details and I think this is mainly because there is only one impl. I agree that we need to clean it up if we want to keep the ability to relatively easy swap storage implementations.

I don't actually much care about that at this point. You could imagine having a mock implementation for some unit tests, but creating even a mock implementation would be quite a lot of work, and most of the unit tests we have are for testing the LayeredRepository implementation details anyway, so a mock implementation of Repository/Timeline wouldn't  help with that.

However, I do think Repository/Timeline provide useful structure and documentation of which parts are public-facing, and which parts are considered internal implementation details, even if we only ever have one implementation. I'm not 100% sure it's a good tradeoff - it does complicate some things, you need to use generics more, etc. - but it does have some documentation value.

> However I like the direction of this PR. It removes circular dependency between Timeline and DatadirTimeline, and makes code that uses both of them cleaner, so I propose to merge these changes and evolve on that further. Maybe during discussions for the RFC we'll come up with better alternative.

Yep, one step at a time. Thanks!

hlinnaka commented on pageserver/src/layered_repository/timeline.rs:
```diff
@@ -634,6 +645,56 @@ impl LayeredTimeline {
+    /// NOTE: counted incrementally, includes ancestors,
+    pub fn get_current_logical_size(&self) -> usize {
+        let current_logical_size = self.current_logical_size.load(AtomicOrdering::Acquire);
+        match usize::try_from(current_logical_size) {
```
If there's a bug in tracking the size, we don't want to throw an error and stop ingesting new WAL because of that. It might make sense to set it to a special value that indicates "broken", though, and recalculate later.

If we start calculating the size lazily in the future, a negative value would be useful to track updates, until the lazy calculation finishes. We'd need another variable or enum to indicate whether the calculation has finished or not though.

LizardWizzard replied:
> we don't want to throw an error and stop ingesting new WAL because of that

Hm, you're right. We can continue ingesting new records after the error. IMO it is still better to emit an error when the erroneous modification happens. It can contain more details, e g type of wal record which lead to the error, maybe a backtrace. With current approach we wont notice until someone calls the api or feedback message is generated (it happens quite often but anyway).

> We'd need another variable or enum to indicate whether the calculation has finished or not though.

Thats true.

hlinnaka replied:
> > we don't want to throw an error and stop ingesting new WAL because of that
> 
> Hm, you're right. We can continue ingesting new records after the error. IMO it is still better to emit an error when the erroneous modification happens. It can contain more details, e g type of wal record which lead to the error, maybe a backtrace. With current approach we wont notice until someone calls the api or feedback message is generated (it happens quite often but anyway).

True, an error in the log when we first detect the problem would be helpful. If the size goes negative, there's a good chance the fault is not with the WAL record where it goes negative, but that we forgot to increment the value at some earlier record. Still, a log message would help to narrow it down.

LizardWizzard replied:
Sort of related: https://github.com/neondatabase/neon/pull/2126#discussion_r930200004 For physical size the reason was in prometheus wrapper. In general I think it would be nice for physical and logical size to agree on what type to use

aome510 commented on pageserver/src/layered_repository/timeline.rs:
```diff
@@ -634,6 +644,58 @@ impl LayeredTimeline {
+            // when ancestor timeline is concurrently updated.
+            //
+            // Logical size 0 means that it was not initialized, so don't believe that.
+            if ancestor_logical_size != 0 && ancestor.get_last_record_lsn() == self.ancestor_lsn {
```
Look like with this change, it goes back to the copy-on-loading approach in https://github.com/neondatabase/neon/pull/2117. One reason we did another PR (https://github.com/neondatabase/neon/pull/2139) was because it might be better to copy-on-branch-creation as we know that the statement `ancestor.get_last_record_lsn() == self.ancestor_lsn` holds during the branch creation. But that's not true when loading the child because parent branch can be modified between.

I'm fine with either approach as both will improve the branch creation time. I still think the one we did in #2139 will bring more gain. 

cc @knizhnik

hlinnaka replied:
Hmm, create_timeline calls `repo.branch_timeline`, and immediately after that, it calls `tenant_mgr::get_local_timeline_with_load`, which calls `load_local_timeline`, which calls `init_logical_size`. Admittedly, there are some semi-expensive operations in between, like writing out the metadata file. But even with the previous coding, if new data is inserted to the parent branch at the same time, the optimization might not work, so this just makes the window slightly larger. I'm ok with that.

In the long run, I think lazy initialization is the way to go. There's no strong reason we couldn't allow the user to use the branch, before the logical timeline size has been calculated.

aome510 replied:
> In the long run, I think lazy initialization is the way to go. There's no strong reason we couldn't allow the user to use the branch, before the logical timeline size has been calculated.

That makes sense. Created https://github.com/neondatabase/neon/issues/2167 for that

aome510 commented on pageserver/src/layered_repository/timeline.rs:
```diff
@@ -269,11 +268,18 @@ pub struct LayeredTimeline {
-    // Initial logical size of timeline (if known).
-    // Logical size can be copied from ancestor timeline when new branch is create at last LSN
-    pub init_logical_size: Option<usize>,
+    /// When did we last calculate the partitioning?
```
I don't get this description. Seeing "when", I expect the variable to be timestamped. Should this be something that mentions "the last partitioning at some LSN"?

hlinnaka replied:
I see your point. I think of LSN as the time dimension here, but you're right, it's not self-evident. This PR just moves the code as is, so I'll try to clarify that as a separate PR.

LizardWizzard commented:
> The point is that there would be trait PostgresDatadirTimeline and also trait MySqlDatadirTimeline.

Gotcha.

> Timeline can implement both, and if you want to store Postgres data in the underlying repository, you call the functions in PostgresDatadirTimeline, and if you want to store MySQL data, you call the functions in MySqlDatadirTimeline.

In this case it would be possible to intermix different datadir flavours in one timeline. E g compiler will be ok if you try to insert mysql data into postgres timeline. So there needs to be some layer of protection. 

For case with different postgres versions it may be desirable if we want to support both versions on the same timeline, or on branches. Though the feature might be interesting (forking your database with different postgres version to try it out) the implementation is really tricky (walredo stuff, upgrade/downgrade page versions back and forth) so I dont think we will have something like that in foreseeable future.

I think this fits the [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) idiom. Mysql stuff is far beyond the horizon. For pg15 support we dont know whether we need separate datadir flavours or support will be implemented differently.

hlinnaka commented:
> > Timeline can implement both, and if you want to store Postgres data in the underlying repository, you call the functions in PostgresDatadirTimeline, and if you want to store MySQL data, you call the functions in MySqlDatadirTimeline.
> 
> In this case it would be possible to intermix different datadir flavours in one timeline. E g compiler will be ok if you try to insert mysql data into postgres timeline. So there needs to be some layer of protection.

Yep. I don't have a good solution in mind for that.

> For case with different postgres versions it may be desirable if we want to support both versions on the same timeline, or on branches. Though the feature might be interesting (forking your database with different postgres version to try it out) the implementation is really tricky (walredo stuff, upgrade/downgrade page versions back and forth) so I dont think we will have something like that in foreseeable future.

Yeah, I wouldn't try to do upgrade on the fly. However, as soon as support upgrading a branch, we'll need to somehow track which LSN-ranges belong to which PostgreSQL version. For example, you would run PostgreSQL v14 up to LSN 1000, and at 1000, you would run pg_upgrade. It would rewrite the catalogs and maybe do some other stuff, and those changes would be written with LSN 1000. From LSN 1000 onwards, you should use v15 to interpret the data. But if you create a branch of that at LSN 900, the new branch would need to use v14. 

> I think this fits the [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) idiom. Mysql stuff is far beyond the horizon. For pg15 support we dont know whether we need separate datadir flavours or support will be implemented differently.

Yep. I'll push this, as it's a step in the right direction.
